### PR TITLE
Deprecation with version support

### PR DESCRIPTION
**Pull request by fantasticfears**
https://meta.discourse.org/t/deprecation-warning-improvement-idea/103241

It's tricky when it comes to Ember part since we have plugin api.

### CODE REVIEW

File: app/assets/javascripts/discourse-common/lib/deprecated.js.es6 [hunk 1]
```diff
@@ -1,3 +1,8 @@
-export default function deprecated(msg) {
-  console.warn(`DEPRECATION: ${msg}`); // eslint-disable-line no-console
+export default function deprecated(msg, opts = {}) {
+  let warning = `DEPRECATION: ${msg}`;
+  let versionInfo = '';
+  if (opts.since) { versionInfo += `It's deprecated since Discourse ${opts.since}.`; }
+  if (opts.dropFrom) { versionInfo += `It will be removed from Discourse ${opts.dropFrom}.`; }
+  if (versionInfo.length > 0) { versionInfo = `\nVersion: ${versionInfo}\n`; }
+  console.warn(`DEPRECATION: ${warning}${versionInfo}`); // eslint-disable-line no-console
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse-common/lib/deprecated.js.es6]: I think you have `DEPRECATION:` twice here. (line 2 & line 7)

Comment (fantasticfears) [reply app/assets/javascripts/discourse-common/lib/deprecated.js.es6]: Fixed.

File: app/assets/javascripts/discourse-common/lib/deprecated.js.es6 [hunk 2]
```diff
@@ -1,3 +1,11 @@
-export default function deprecated(msg) {
-  console.warn(`DEPRECATION: ${msg}`); // eslint-disable-line no-console
+export default function deprecated(msg, opts = {}) {
+  msg = ['Deprecation notice:', msg];
+  if (opts.since) { msg.push(`(deprecated since Discourse ${opts.since})`; }
+  if (opts.dropFrom) { msg.push(`(removal in Discourse ${opts.dropFrom})`; }
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse-common/lib/deprecated.js.es6]: Missing a close bracket here, and on the line above.

Comment (fantasticfears) [reply app/assets/javascripts/discourse-common/lib/deprecated.js.es6]: Fixed

**Comment by davidtaylorhq** [comment]
Merged, thanks @fantasticfears

**Pull request merged**